hail/python/hailtop/hailctl/batch/submit.py

```diff
@@ -1,108 +1,185 @@
+from hailtop.aiotools.router_fs import RouterAsyncFS
+from hailtop.config import (
+    get_deploy_config,
+    get_hail_config_path,
+)
+
+from .batch_cli_utils import StructuredFormatPlusTextOption
+
+
+class HailctlBatchSubmitError(Exception):
+    def __init__(self, message: str, exit_code: int):
+        self.message = message
+        self.exit_code = exit_code
+
+
+async def submit(
+    image: str,
+    entrypoint: List[str],
+    name: Optional[str],
+    cpu: str,
+    memory: str,
+    storage: str,
+    machine_type: Optional[str],
+    spot: bool,
+    workdir: str,
+    cloudfuse: Optional[List[Tuple[str, str, bool]]],
+    env: Optional[Dict[str, str]],
+    billing_project: Optional[str],
+    remote_tmpdir: Optional[str],
+    regions: Optional[List[str]],
+    requester_pays_project: Optional[str],
+    attributes: Optional[Dict[str, str]],
+    volume_mounts: Optional[List[Tuple[str, str]]],
+    shell: Optional[str],
+    output: StructuredFormatPlusTextOption,
+    wait: bool,
+    quiet: bool,
+    # FIXME: requester pays config
+):
+    async with AsyncExitStack() as exitstack:
+        fs = RouterAsyncFS()
+        exitstack.push_async_callback(fs.close)
+
+        backend = hb.ServiceBackend(billing_project=billing_project,
+                                    remote_tmpdir=remote_tmpdir,
+                                    regions=regions,
+                                    gcs_requester_pays_configuration=requester_pays_project)
+
+        exitstack.push_async_callback(backend.async_close)
+
+        local_user_config_dir = '/hail-config/'
+
+        b = hb.Batch(backend=backend, name=name, attributes=attributes)
+
+        config_file_paths = await get_user_config_files(fs)
+        config_file_inputs = [(os.path.basename(path), b.read_input(path)) for path in config_file_paths]
+        config_file_str = "\n".join(f'mv {input} {local_user_config_dir}{file_name}' for file_name, input in config_file_inputs)
+
+        volume_mount_inputs = []
+        mkdirs_needed = {local_user_config_dir}
+
+        for src, dest in volume_mounts:
+            if await fs.isfile(src):
+                mkdirs_needed.add(os.path.dirname(dest))
+            else:
+                if src.endswith('/') and not dest.endswith('/'):
+                    raise ValueError('copy and renaming a directory is not supported.')
+                if not src.endswith('/') and dest.endswith('/'):
+                    dest = dest.rstrip('/') + '/' + os.path.basename(src)
+
+                dest = dest.rstrip('/') + '/'
+                mkdirs_needed.add(dest)
+
+            volume_mount_paths = await get_volume_mount_files(fs, src)
+
+            for path in volume_mount_paths:
+                path_relname = os.path.relpath(path, src)
+                if path_relname == '.':
+                    path_relname = ''
+                mkdirs_needed.add(os.path.dirname(f'{dest}{path_relname}').rstrip('/') + '/')
+
+                if dest.endswith('/'):
```

Comment:
I don't think you need the branch here. Two lines up we've already used `f'{dest}{path_relname}'` in all cases (and it took me a bit to see why that's always okay!).